massifs/mmriver.go

```diff
@@ -0,0 +1,246 @@
+	"github.com/fxamacker/cbor/v2"
+
+	commoncose "github.com/datatrails/go-datatrails-common/cose"
+	"github.com/datatrails/go-datatrails-common/logger"
+	"github.com/datatrails/go-datatrails-merklelog/mmr"
+)
+
+// MMRIVER COSE Receipts to accompany our COSE MMRIVER seals
+
+type MMRiverInclusionProof struct {
+	Index         uint64   `cbor:"1,keyasint"`
+	InclusionPath [][]byte `cbor:"2,keyasint"`
+}
+
+type MMRiverConsistencyProof struct {
+	TreeSize1        uint64   `cbor:"1,keyasint"`
+	TreeSize2        uint64   `cbor:"2,keyasint"`
+	ConsistencyPaths [][]byte `cbor:"3,keyasint"`
+	RightPeaks       [][]byte `cbor:"4,keyasint"`
+}
+
+type MMRiverVerifiableProofs struct {
+	InclusionProofs   []MMRiverInclusionProof   `cbor:"-1,keyasint,omitempty"`
+	ConsistencyProofs []MMRiverConsistencyProof `cbor:"-2,keyasint,omitempty"`
+}
+
+// MMRiverInclusionProofHeader provides for encoding, and defered decoding, of
+// COSE_Sign1 message headers for MMRIVER receipts
+type MMRiverVerifiableProofsHeader struct {
+	VerifiableProofs MMRiverVerifiableProofs `cbor:"396,keyasint"`
+}
+
+/*
+func SetMMRiverInclusionProofsHeader(
+	msg *commoncose.CoseSign1Message, massif MassifReader, mmrSize, mmrIndex uint64) error {
+	msg.Headers.Unprotected[VDSCoseReceiptProofsTag] = proofs
+}*/
+
+// VerifySignedInclusionReceipts verifies a signed COSE receipt encoded according to the MMRIVER VDS
+// on success the produced root is returned.
+// Signature verification failure is not an error, but the returned root will be nil and the result will be false.
+// All other unexpected issues are returned as errors, with a false result and nil root.
+// Note that MMRIVER receipts allow for multiple inclusion proofs to be attached to the receipt.
+// This function returns true only if ALL receipts verify
+func VerifySignedInclusionReceipts(
+	ctx context.Context,
+	receipt *commoncose.CoseSign1Message,
+	candidates [][]byte,
+) (bool, []byte, error) {
+
+	var err error
+
+	// ignore any existing payload
+	receipt.Payload = nil
+
+	// We must return false if there are no candidates
+	if len(candidates) == 0 {
+		return false, nil, fmt.Errorf("no candidates provided")
+	}
+
+	var header MMRiverVerifiableProofsHeader
+	err = cbor.Unmarshal(receipt.Headers.RawUnprotected, &header)
+	if err != nil {
+		return false, nil, fmt.Errorf("MMRIVER receipt proofs malformed")
+	}
+	verifiableProofs := header.VerifiableProofs
+	if len(verifiableProofs.InclusionProofs) == 0 {
+		return false, nil, fmt.Errorf("MMRIVER receipt inclusion proofs not present")
+	}
+
+	// permit *fewer* candidates than proofs, but not more
+	if len(candidates) > len(verifiableProofs.InclusionProofs) {
+		return false, nil, fmt.Errorf("MMRIVER receipt more candidates than proofs")
+	}
+
+	var proof MMRiverInclusionProof
+
+	proof = verifiableProofs.InclusionProofs[0]
+	receipt.Payload = mmr.IncludedRoot(
+		sha256.New(),
+		proof.Index, candidates[0],
+		proof.InclusionPath)
+
+	err = receipt.VerifyWithCWTPublicKey(nil)
```

Comment:
minor: why not verify the signature of the receipt before checking the inclusion root of the first candidate?

Reply:
ahh because we need to set the payload in order to get the correct pre-image for the signature to verify. 

I get it!